### PR TITLE
RUBY-2426: Check if indexes for GridFS exist prior to creating them

### PR DESCRIPTION
**Pull request by alexbevi**
https://jira.mongodb.org/browse/RUBY-2426

### CODE REVIEW

**Comment by alexbevi** [comment]
At least one spec failure with this PR that needs cleaning up:
```
MONGODB_URI="mongodb://dev:dev@localhost:27017/" rake spec SPEC=spec/mongo/grid/fs_bucket_spec.rb
                                                                                                                                                                                    
  1) Mongo::Grid::FSBucket#insert_one when the index creation encounters an error raises the error to the user
     Failure/Error:
       expect {
         fs.insert_one(file)
       }.to raise_error(Mongo::Error::OperationFailure)
     
       expected Mongo::Error::OperationFailure, got #<RSpec::Mocks::MockExpectationError: (#<Mongo::Collection:0x8280 namespace=ruby-driver.fs.files>).indexes(no args)
           expected: 0 times with any arguments
           received: 1 time> with backtrace:
         # ./lib/mongo/grid/fs_bucket.rb:491:in `ensure_indexes!'
         # ./lib/mongo/grid/fs_bucket.rb:167:in `insert_one'
         # ./spec/mongo/grid/fs_bucket_spec.rb:397:in `block (5 levels) in <top (required)>'
         # ./spec/mongo/grid/fs_bucket_spec.rb:396:in `block (4 levels) in <top (required)>'
         # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
         # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `loop'
         # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
         # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
         # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
         # ./spec/support/background_thread_registry.rb:62:in `block (2 levels) in <top (required)>'
     # ./spec/mongo/grid/fs_bucket_spec.rb:396:in `block (4 levels) in <top (required)>'
     # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `loop'
     # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /home/alex/.rvm/gems/ruby-2.7.1/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
     # ./spec/support/background_thread_registry.rb:62:in `block (2 levels) in <top (required)>'
```

File: lib/mongo/grid/fs_bucket.rb
Comment (alcaeus) [inline lib/mongo/grid/fs_bucket.rb:505]: I'll note that the "Enumerating indexes" spec [contains a suggested implementation](https://github.com/mongodb/specifications/blob/master/source/enumerate-indexes.rst#enumeration-getting-index-information) on how to enumerate indexes. That implementation checks for the `NamespaceNotFound` error and returns an empty array in that case.

If the driver should continue raising an exception when listing indexes on non-existing collections, this method should catch the exception itself and return `false` in that case. This also removes the need for running two commands for an index check.

**Review by alcaeus — COMMENTED** [review]
I'll let @p-mongo decide on how to best handle the `NamespaceNotFound` error (either in the `listIndexes` API or in the helper), but the logic can be optimised either way.

File: lib/mongo/grid/fs_bucket.rb
Comment (p-mongo) [reply lib/mongo/grid/fs_bucket.rb:505]: +1

**Comment by p-mongo** [comment]
I believe there needs to be a test added that the driver is able to write to gridfs when the user lacks permission to create indexes. This would, I expect, involve defining another test user.

The spec language is somewhat confusing in this regard:

> Drivers MUST check whether the indexes already exist before attempting to create them. This supports the scenario where an application is running with read-only authorizations.

Surely, an application "running with read-only authorizations" won't be able to write files to gridfs either. Does the spec indeed mean an environment where inserts are allowed but index creation is not?

**Comment by p-mongo** [comment]
The documentation should also be updated. https://docs.mongodb.com/ruby-driver/master/tutorials/ruby-driver-gridfs/ currently says:

> Note that the first time a file is inserted, it will create the required index for you on the chunks collection.

One might read this to mean "the first time a file is inserted *using the Ruby driver*" which isn't accurate; additionally, the files index isn't mentioned and the lack of permissions isn't addressed either. I suggest something like the following wording instead:

> When the driver is inserting the first file into a bucket, it will attempt to create the required indexes on `files` and `chunks` collections. The required indexes are as follows: (both index specifications) 
> If the indexes cannot be created, such as due to the current user lacking the permissions to do so, the file insert will be aborted.
> If the bucket already has files, the driver will not attempt to create indexes, even if they are missing and the current user has permissions to create them. In this case a database administrator should create the needed indexes as soon as possible to ensure data integrity.

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb]: Rewrote the logic here

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [inline lib/mongo/grid/fs_bucket.rb:509]: @alcaeus I think this implementation is more aligned with what you had in mind

Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb:509]: > I believe there needs to be a test added that the driver is able to write to gridfs when the user lacks permission to create indexes. This would, I expect, involve defining another test user.

@p do you think this test should be added?

**Comment by p-mongo** [comment]
Yes I think so.

File: lib/mongo/grid/fs_bucket.rb
Comment (p-mongo) [inline lib/mongo/grid/fs_bucket.rb:501]: This appears to create the index if it exists?

File: lib/mongo/grid/fs_bucket.rb
Comment (p-mongo) [inline lib/mongo/grid/fs_bucket.rb:494]: This seems to create each index based on the presence of data in the respective backing collection. I had a different understanding of the mechanism so the docs should be updated to match the actual behavior.

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb:501]: 🤦🏻

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb:494]: I'm going based on the [spec's "Before write operations"](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#id19) guidance:

```
Immediately before the first write operation on an instance of a GridFSBucket class is attempted (and not earlier), drivers MUST:

determine if the files collection is empty using the primary read preference mode.
and if so, create the indexes described above if they do not already exist
To determine whether the files collection is empty drivers SHOULD execute the equivalent of the following shell command:

> db.fs.files.findOne({}, { _id : 1 })
If no document is returned the files collection is empty.
```

**Comment by alexbevi** [comment]
@p / @comandeo finally got around to this. hopefully this is good now, though might need rebasing :/

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb:494]: I haven't addressed this yet, so may still have to revise the docs before this can be merged

File: lib/mongo/grid/fs_bucket.rb
Comment (alexbevi) [reply lib/mongo/grid/fs_bucket.rb:501]: Fixed my backwards logic (finally)

**Review by p-mongo — APPROVED** [review]

**Review by comandeo — APPROVED** [review]

**Pull request merged**
